### DISCUSSION TITLE
Any advice for smoother Matrix animation?

### DISCUSSION THREAD

**Discussion by RadhiFadlillah** [charmbracelet/bubbletea#1052] [Q&A]
Hi everyone :wave: 

I'm trying to make Matrix animation like the one from `cmatrix` command. I'm succeed to make it [here](https://github.com/RadhiFadlillah/charm-matrix) but its performance leave a lot to be desired.

I noticed my CPU usage become very high and my PC feels lagging. In Konsole, it run kinda smooth at first, but the longer it run the more it stutters:

[Charm Matrix in Konsole](https://github.com/charmbracelet/bubbletea/assets/6129042/4f4c0473-ec1d-4636-b98d-b0d3c479da7f)

In VS Code integrated terminal, the animation flickers a lot:

[Charm Matrix in VS Code](https://github.com/charmbracelet/bubbletea/assets/6129042/cb1890fa-aafa-4bdf-a1e1-c8a50f963012)

I know there is a new renderer [planned](https://github.com/charmbracelet/bubbletea/issues/1032#issuecomment-2143488845) to fix flicker issue, but I wonder if there are something I can do on my side. The only thing that I did so far is to use `strings.Builder` instead of concatenating string like `s += "A" + "b"`.

Thanks!

**Comment by Gaurav-Gosain** [comment]
https://github.com/charmbracelet/bubbletea/assets/55647468/6c01047d-fa57-4814-aeb0-d940a1e6fe73

Runs pretty smooth for me without any jitters or excessive cpu/ram usage :thinking: had the thought at first it might be because of the terminal I'm using (kitty, which is GPU accelerated) but then verified it runs fine on vscode too :man_shrugging:

**Reply by RadhiFadlillah** [reply]
May I know what is your system spec? Mine is:

```
CPU: Intel i7-8550U (8) @ 4.000GHz 
GPU: Intel UHD Graphics 620 
RAM: 16GB
```

I've tried a GPU accelerated terminal `wezterm`, the CPU usage goes down but the flicker still occurs.

I also wonder if display manager could improve the performance, so I tried using Wayland and X but there are almost no differences in either of them.

**Reply by Gaurav-Gosain** [reply]
I've got a `Intel(R) Core(TM) i7-10700` and a `GeForce RTX 3070`, am using Wayland.

**Reply by meowgorithm** [reply]
Just a related note here: we're working internally on some pretty major renderer improvements to Bubble Tea to improve performance and reduce the amount of data transfer between Bubble Tea and the terminal. No ETA yet, but it's coming along and feeling good.

**Comment by meowgorithm** [comment]
This looks SO awesome. In terms of performance: we're currently working on an update to the Bubble Tea renderer that should improve performance significantly.

**Reply by RadhiFadlillah** [reply]
Thanks for the compliment, and double thanks for making Bubble Tea :)

Yes, I've seen you mentioned in several issues about the next renderer. Very excited and looking forward to it :+1:
